### PR TITLE
Remove blank lines at start or end of block 

### LINKED ISSUE #1274: Remove unnecessary new lines at start or end of block
For example:

```rust
fn main() {




    let x = 1;




}
```

Could be formatted to:

```rust
fn main() {
    let x = 1;
}
```

### PR DESCRIPTION
**Pull request by sinkuu**
Fixes #1274

### CODE REVIEW

File: src/visitor.rs [hunk 1]
```diff
@@ -131,6 +131,27 @@ impl<'a> FmtVisitor<'a> {
         self.block_indent = self.block_indent.block_indent(self.config);
         self.buffer.push_str("{");
 
+        if self.config.remove_blank_lines_at_start_or_end_of_block() {
+            if let Some(stmt) = b.stmts.first() {
+                let snippet = self.snippet(mk_sp(self.last_pos, stmt.span.lo));
```
Comment (topecongiro) [inline src/visitor.rs]: Could we use a byte pos from inner attributes to construct this snippet and avoid calling `s.trim_left_matches(&[' ', '\t', '\r', '\n'][..]).len()]` below?

Comment (sinkuu) [inline src/visitor.rs]: Done.

File: src/visitor.rs [hunk 2]
```diff
@@ -132,22 +133,44 @@ impl<'a> FmtVisitor<'a> {
...
+                let attr_lo = inner_attrs
+                    .and_then(|attrs| {
+                        attrs
+                            .iter()
+                            .filter(|a| a.style == ast::AttrStyle::Inner)
+                            .nth(0)
+                            .map(|attr| attr.span.lo)
+                    })
+                    .or_else(|| {
+                        // syntax BUG: `attrs()` returns nothing for items
```
Comment (sinkuu) [inline src/visitor.rs]: https://github.com/rust-lang/rust/pull/43807

File: src/visitor.rs [hunk 3]
```diff
@@ -132,22 +133,44 @@ impl<'a> FmtVisitor<'a> {
...
-                            // There may be inner attributes
-                            let s = &s[..s.len() -
-                                           s.trim_left_matches(&[' ', '\t', '\r', '\n'][..]).len()];
-                            s.rfind('\n')
+            if let Some(first_stmt) = b.stmts.first() {
+                let attr_lo = inner_attrs
+                    .and_then(|attrs| {
+                        attrs
+                            .iter()
+                            .filter(|a| a.style == ast::AttrStyle::Inner)
```
Comment (topecongiro) [inline src/visitor.rs]: We could use `inner_attributes()` in utils.rs here.

File: src/visitor.rs (hunk 2 above)
Comment (topecongiro) [inline src/visitor.rs]: I think we could avoid this by using `Spanned` trait in lib.rs (`first_stmt.span().lo`).

**Comment by topecongiro** [comment]
Thank you for you update! I left inline comments again, please feel free to ask if there are anything unclear.

File: src/visitor.rs (hunk 2 above)
Comment (sinkuu) [inline src/visitor.rs]: I tried, but test failed within `tests/target/attrib-extern-crate.rs` in the same way as `first_stmt.attrs()`.

https://github.com/rust-lang-nursery/rustfmt/blob/0ee76bec0f6a76b7b26eebdbd5700e9c727a96b3/src/lib.rs#L112-L118 looks like it does nothing to `StmtKind::Item`. Should this be modified to include attributes for items?

Comment (sinkuu) [inline src/visitor.rs]: It turned out that the behavior of `attrs()` is intentional; attributes for an item in a statement position don't belong to the statement (https://github.com/rust-lang/rust/commit/b968ee36564abd96da72c381f49bf8009f4b1870). Perhaps `Spanned` follows the convention.

**Comment by topecongiro** [comment]
Thanks!

**Pull request merged**